### ISSUE TITLE
@BeforeExample issues for 2nd and subsequent tests

### ISSUE THREAD

**Issue by nigelcharman** [concordion/concordion#262]
Reported by `Vladimir100` to the Concordion google group.

> We found a serious bug in Concordion 2.1.1 and 2.1.2 that prevents us from switching to the latest version. However, it works for 2.0.3.
>
> Here is the md file.
>
```markdown
## Sum and Divide Example

### [Example 1. 2 + 4](- "sum2-4")
The sum of two numbers [2] (- "#firstNumber")
   and  [4] (- "#secondNumber") will be [6] (- "?=getSum(#firstNumber, #secondNumber)").

### [Example 2. 4 / 2](- "divide4-2")  
Division of two numbers [4] (- "#firstNumber")
   and  [2] (- "#secondNumber") will be [2] (- "?=getQuotient(#firstNumber, #secondNumber)"). 
```
>
> Here is the fixture.
>
```java
@RunWith(ConcordionRunner.class)
public class AssertEq
{
 private String VBTest;
 
 @BeforeExample
 public void init()
 {
 VBTest = "Tony";
 }
 
 public int getSum(int a, int b)
 {
 System.out.println(VBTest);
 return a + b;
 }
 
 public int getQuotient(int a, int b )
 {
 System.out.println(VBTest);
 return a/b;
 }
}
```
Here is the outcome:
```
Tony

file:///var/folders/zp/74sv30dj3k5422sp02rcr6v00000gr/T/concordion/com/bluescape/VB/tests/AssertEq.html#sum2-4
Successes: 1, Failures: 0
null

file:///var/folders/zp/74sv30dj3k5422sp02rcr6v00000gr/T/concordion/com/bluescape/VB/tests/AssertEq.html#divide4-2
Successes: 1, Failures: 0

file:///var/folders/zp/74sv30dj3k5422sp02rcr6v00000gr/T/concordion/com/bluescape/VB/tests/AssertEq.html
Successes: 2, Failures: 0
```
As you can see, the variable VBTest is not initialized as "Tony" in the second example. 
It is shown as null.

**Comment by nigelcharman** [comment]
This is related to #246 and is a knock-on effect of the changes from #225.

The before and after example listeners are operating on the wrong fixture instance, since they are initialised at build-time in `ConcordionBuilder` before the individual example fixture instances are created at runtime in `ConcordionRunner.invoke(ConcordionFrameworkMethod, Object)`. This means that when the `@BeforeExample` method is called, it is operating on the wrong fixture instance.

The `ConcordionScoped` test case checks that "before examples" work correctly with field initialisation. However the "before examples" work differently to `@BeforeExample` code. Changing this test case to use `@BeforeExample` illustrates this issue.

As experiments the following workarounds fix this issue:
1. Moving the call to `invokeMethods(...` from `FixtureInstance.beforeExample(String)` to `FixtureInstance.beforeProcessExample(String)` (however this would undo #225).
2. Adding to `XMLSpecification`:
line 14:
`public static final ThreadLocal<Fixture> FIXTURE_HOLDER = new ThreadLocal<Fixture>();`
line 87:
`                FIXTURE_HOLDER.set(fixture);`
and to `ExampleCommand`:
line 39:
`            listeners = new ArrayList<ExampleListener>();`
`            Fixture fixture = XMLSpecification.FIXTURE_HOLDER.get();`
`            addExampleListener(new FixtureExampleHook(fixture));`
_This is probably the ugliest hack I've ever done, I include it here only for reference while we research a clean solution!!! At least I didn't commit it. Please don't hold it against me :)_

**Comment by nigelcharman** [comment]
Fixed in #264

**Comment by nigelcharman** [comment]
Fixed in 2.2.0
